037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+# Automatic rebalance on cluster scaling
+
+This feature is about enabling the cluster operator to get an optimization proposal (to be auto or manually approved) and run the corresponding rebalancing when the user scales the cluster up or down by changing the `spec.kafka.replicas` field on the `Kafka` custom resource.
+The scaling operation is then subordinated to the rebalancing operation and if it is approved and runs successfully.
+
+## Current situation
+
+Currently, there is no automatic rebalancing when the user scales the cluster up or down.
+It means that:
+
+* after scaling up the cluster, the user has to create a `KafkaRebalance` custom resource with the `spec.mode: add-brokers` and the `spec.brokers` with the added brokers, in order to rebalance the cluster and moving some replicas to the newly added brokers.
+* before scaling down the cluster, the user has to create a `KafkaRebalance` custom resource with the `spec.mode: remove-brokers` and the `spec.brokers` with the brokers to remove, in order to rebalance the cluster and moving replicas out of the brokers to be removed.
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two step process.
+
+* scaling up the cluster, then running a rebalance.
+* running a rebalance, then scaling down the cluster.
+
+In order to simplify, it would be great for the user to deal with just scaling the cluster by changing the `spec.kafka.replicas` field and then having the operator taking care to run a rebalancing operation automatically if requested.
+
+## Proposal
+
+We would need to extend the `Kafka` custom resource to allow the user specifying that he wants auto-rebalancing on scaling.
```

Comment:
I agree, I think an annotation would be more flexible at this stage.

Reply:
If I got it correctly and IIRC ... the `KafkaConnect` used to have a `strimzi.io/use-connector-resources: "true"` annotation to enable the usage of `KafkaConnector` CRs. In this case the API is more complex, it's not just about enabling the auto-rebalancing but specifying modes and corresponding templates.

Reply:
Well, you can stick everything into an annotation as a JSON for example. It might help to see what is the best way going forward ... a separate resource, fields inside the Kafka resource etc.

Reply:
But in general, I might have not added here if I read the whole proposal first ... this part talks only about the mode basically, which would be much easier. But if you get back to the discussions about node pools, there might be a lot of other changes to the Kafka CRD and maybe long term it might fit also somewhere else. E.g. should the whole CC staff move to a separate CRD which would give it more space for this kind of configuration?

Reply:
I know. Lot of questions and not too many answers :-D

Reply:
That was my fear, a big single JSON annotation.
Based on my latest refactoring of the auto-rebalancing configuration, it should be an `autoRebalance` annotation containing the array modes with corresponding templates. 

```JSON
autoApproval: "[
   {
      "mode":[
         "add-brokers",
         "remove-brokers"
      ],
      "template":"my-add-brokers-rebalancing-template"
   },
   {
      "mode":[
         "full"
      ],
      "template":"my-full-rebalancing-template"
   }
]"
```

Reply:
Do you really want to have the mode as an array? Validation and error handling will be harder IMHO.

Reply:
Don't cross discussions please. This is about putting in a potential annotation, can you make the same comment on the proposal directly?

Reply:
No, I commented here, now you know about it. You can incorporate it or not depending on what you think about it.

Reply:
Regardless of where to comment, if was a good idea so that `mode` could be an enum and we can have validation at Kubernetes API level. I made the change in the proposal.
Let's see here if we reach an overall agreement to use an annotation instead of CRD changes since the beginning.